bytecomp/matching.ml

```diff
@@ -2225,163 +2568,190 @@ let mk_failaction_pos partial seen ctx defs  =
     None,fail,jmps
   end else begin (* Too many non-matched constructors -> reduced information *)
     if dbg then eprintf "POS->NEG!!!\n%!" ;
-    let fail,jumps =  mk_failaction_neg partial ctx defs in
+    let fail,jumps =  mk_failaction_neg loc partial ctx defs in
     if dbg then
       eprintf "FAIL: %s\n"
         (match fail with
         | None -> "<none>"
-        | Some lam -> string_of_lam lam) ;
+        | Some block -> string_of_block block) ;
     fail,[],jumps
   end
 
-let combine_constant loc arg cst partial ctx def
-    (const_lambda_list, total, _pats) =
-  let fail, local_jumps =
-    mk_failaction_neg partial ctx def in
+type combined = Lambda.block * jumps_union
```

Comment:
This type synonym doesn't make much sense either; `Lambda.block * jumps` is the return type of almost all compilation functions in the module, not specific to this `combine` operation.

Reply:
I don't understand why that makes the type synonym not make sense.  You (or @trefis) will need to suggest a better name, I'm afraid.

Reply:
The reason `combined` does not make sense is that it names the values from the transformation that produced them, and not their structure. This is as if I had defined

```ocaml
type added = int
type multiplied = int
val (+) : int -> int -> added
val ( * ) : int -> int -> multiplied
```
and then I started using `added` or `multiplied` randomly in the rest of my program, depending on whether the value flowing to that point have been computed last from an addition or a multiplication.

I would suggest to use `Lambda.block * jumps` like all other optimization functions in the file. I don't know of a better name. (Many functions in the module are actually using a state monad of the form `jumps -> Lambda.block * jumps`, and that may be summarized as `Lambda.block compiler`, but doing this properly would require small changes all over the code and it's best left out of the scope of this PR.)

Reply:
That's not a good general rule, in my opinion.  However I will substitute it out since you insist.

Reply:
I'm open to a discussion on why you think defining `combined` as you did above is a good idea; right now I don't know why you think it's a good idea.

(My mental model is that you find it helpful to add type annotations as you let type-checking errors drive you to the change, and that you picked names that are inconsistent with the intent of the code or rest of the file because you were only looking at each breakage locally. I see nothing terribly wrong with that, as long as the names have disappeared or become meaningful and consistent by the time the code is upstreamed.)

If you don't wish to argue further on the topic of `combined`, then yes I would be in favor of removing the synonym.